pandas/core/indexes/datetimes.py

```diff
@@ -30,6 +32,29 @@
 from pandas.core.indexes.extension import inherit_names
 from pandas.core.tools.times import to_time
 
+if TYPE_CHECKING:
+    from pandas import DataFrame, PeriodIndex, TimedeltaIndex
+
+
+def dispatch_to_array_and_wrap(delegate) -> Callable[[F], F]:
```

Comment:
not the final location for this. may do PeriodIndex next and move to share.

duplication with inherit_from_data. different in that inherit_from_data creates a method, whereas this wraps a method.

it maybe that we eventually replace all uses inherit_from_data and therefore can delete rather than deduplicate.